### PR TITLE
Port allmydata.test.no_network to Python 3

### PR DESCRIPTION
**Pull request by itamarst**
Originally was going to be `test_encode.py`, thus the branch name, but pivoted to a dependency.

Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3416

### CODE REVIEW

**Review by sajith — APPROVED** [review]
This is stacked on top of #811, correct?  I'm sorry I dropped the ball on that one. :-(

Looks good to me, minor questions/quibbles apart.  Please go ahead and merge when this is ready.

File: src/allmydata/immutable/upload.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
      DEFAULT_MAX_SEGMENT_SIZE, IProgress, IPeerSelector
 from allmydata.immutable import layout
 
-from six.moves import cStringIO as StringIO
+from io import BytesIO
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/immutable/upload.py:30]: StringIO is (unicode) strings on Python 3.

File: src/allmydata/node.py [hunk 2]
```diff
@@ -362,7 +362,7 @@ def get_or_create_private_config(self, name, default=_None):
             if default is _None:
                 raise MissingConfigEntry("The required configuration file %s is missing."
                                          % (quote_output(privname),))
-            if isinstance(default, basestring):
+            if isinstance(default, (bytes, unicode)):
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/node.py:365]: This doesn't change the semantics, basestring is bytes + unicode in Python 2.

File: src/allmydata/test/common_py3.py [hunk 3]
```diff
@@ -123,7 +123,7 @@ def done(res):
 class ReallyEqualMixin(object):
     def failUnlessReallyEqual(self, a, b, msg=None):
         self.assertEqual(a, b, msg)
-        self.assertEqual(type(a), type(b), "a :: %r, b :: %r, %r" % (a, b, msg))
+        self.assertEqual(type(a), type(b), "a :: %r (%s), b :: %r (%s), %r" % (a, type(a), b, type(b), msg))
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/common_py3.py:126]: It's a mixin for `TestCase` classes, so it assumes it's there.

File: src/allmydata/test/no_network.py [hunk 4]
```diff
@@ -308,6 +325,7 @@ def __init__(self, basedir, num_clients, num_servers,
             d.addCallback(lambda c: self.clients.append(c))
 
             def _bad(f):
+                print(f)
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/no_network.py:328]: Well... it's very useful to not have silent failures in tests. But maybe it should be a log message.

[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/test/no_network.py:328]: Maybe even a `log.err`?

[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/test/no_network.py:328]: Although I see that `_setup_errors` _may_ get inspected later and if anything is there, an exception raised.

[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/no_network.py:328]: I'll just delete it.

File: src/allmydata/client.py [hunk 5]
```diff
@@ -739,12 +739,12 @@ def init_node_key(self):
...
         def _make_key():
             private_key, _ = ed25519.create_signing_keypair()
-            return ed25519.string_from_signing_key(private_key) + "\n"
+            return ed25519.string_from_signing_key(private_key) + b"\n"
 
         private_key_str = self.config.get_or_create_private_config("node.privkey", _make_key)
         private_key, public_key = ed25519.signing_keypair_from_string(private_key_str)
         public_key_str = ed25519.string_from_verifying_key(public_key)
-        self.config.write_config_file("node.pubkey", public_key_str + "\n", "w")
+        self.config.write_config_file("node.pubkey", public_key_str + b"\n", "wb")
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/client.py:747]: On Python 3, "w" means Unicode, "wb" means bytes.

**Pull request merged**